### ISSUE TITLE
Malicious CT file can result in CR LF Injection and Stored-XSS on all endpoints with the targeted extension.

### ISSUE THREAD

**Issue by DEMON1A** [Praudyogikee/SimplyJServer#6]
- Hi there, during some cool digging today i found that CR LF injection is possible on SimplyJServer when the user is using an external untrusted CT file. as you can see in the code below:

```java
content_types = TextToHashmap.Convert(db, ",", ":");
```

- You're using the `TextToHashmap` Functuion on your code to split two things, the `,` to parse every extension with it's content type. then using another split to seprate the file extension from it's content type. then store them on `content_types` variable. the problem here is that. you're trusting the user provided content_types to add to the server response without any validation. 

```java
public static void write(DataOutputStream s, byte[] res, String content, String S) {
			try {
				s.write((S + "\r\n").getBytes());
				s.write("Server: SimplyJServer\r\n".getBytes());
				s.write(("Content-Length: " + res.length + "\r\n").getBytes());
				s.write("Connection: Keep-Alive\r\n".getBytes());
				// s.write("Content-Encoding: gzip\r\n".getBytes());
				s.write("Keep-Alive: timeout=5, max=1000\r\n".getBytes());
				if (content.equals("text/html")) {
					s.write(("Content-Type: " + content + ";charset=UTF-8\r\n\r\n").getBytes());
				} else {
					s.write(("Content-Type: " + content + "\r\n\r\n").getBytes());
				}
				// byte[] temp;
				// int i = 0;
				// res = compress(res);
				/*
				 * unComment all ONLY when you want to use GZIP
				 */
				s.write(res);
				s.flush();
				s.close();
			} catch (Exception e) {
				
			}
		}
``` 

- So since the split functions only uses these two variables: `,` and `:`. the attacker is able to inject a `CR LF` character on the content type resulting in a new lines getting added into the server response. that includes headers and content. but because you're splitting: `:` character. it's not really possible to add headers into the response since the header value will get splitted and only the header name will be returned. but the attacker is able to add multiple new lines to add it's own content resulting in mass Stored-XSS on `/*.extension`.

- I created a simple CT file. that causes the Stored-XSS when `.test` extension is requested. here's the `.ct`:

```
.test:text/html


<h1>HelloWorld</h1>>,.3dm:x-world/x-3dmf,.3dmf:x-world/x-3dmf,.a:application/octet-stream
```

- Use it with `ct` argument. then visit any `.test` extension. and the response should be a rendered H1 on the page.

```
[root@demonia]:~ - curl "http://localhost:8888/file.test"

<h1>HelloWorld</h
```

**Comment by Zelakolase** [comment]
Hi There,
We don't recommend users to use untrusted CT file, and only rare cases where we can customise CT.

Also, checking through a value would lower the performance. Improvement isn't needed since this case doesn't happen a lot!

**Comment by DEMON1A** [comment]
Can i ask how would `content.replaceAll('\n')` on the response lower the performance? spreading such a CT config can result in random attacks on some servers. also, your application should be escaping `\n` as well. or that will result in a server errors on the response if the user gonna add a new content type on a new line. and i can see it's not documented that the CT shouldn't contain new lines. 

Also, CT is user controlled content and there's an argument for it too. that gives a high chance for the exploit to happen.

**Comment by Zelakolase** [comment]
Hi there,

String utilities such as 'replace','replaceAll','concat','split' are usually slow especially on large strings. It would be a sacrifice to do such thing, however I'll try to test it.

Also, we've decided to make another version of SimplyJServer to be more dynamic and fast, the new version will be released anytime between April 30 and August 30. Make sure you follow t.me/PraudNews to monitor changes.

You're partially correct, '\n' is an abnormal thing to exist in CT. However, I can't simulate a scenario so CT can be controlled to exploit it.

Yes, CT is a user controlled content. Any exploit that would exploit it using command arguments means that the hacker is already on the system. There's no logic when it comes to accessing the system to just change the CT. This would be a hilarious attack not a real one.

When you develop an application that writes on files. Make sure that the range of operations in '/public_html', otherwise such attack can happen.
We'll review the 'writeFile' function and get back to you.

**Comment by Zelakolase** [comment]
Hey again,

I just reviewed the 'Filenamer' function. 'writeFile' checks if the file is in 'public_html' folder or not by taking a Boolean to pass it through 'Filenamer' function to see the appropriate path. Based on a previous issue, we added a filter for a vulnerability.

We encourage developers to do I/O operations inside 'public_html'. Otherwise, such attack can happen.

If you want to share your thoughts about a real life scenario exploiting this vulnerability, that would be great.
